sdk/conftest.py

```diff
@@ -55,4 +70,125 @@ def clean_cached_resources():
+
+    For more details and usage examples, refer to
+    https://github.com/Azure/azure-sdk-for-python/blob/main/doc/dev/test_proxy_migration_guide.md.
+    """
+    test_id, recording_id, variables = start_proxy_session
+    original_transport_func = RequestsTransport.send
+
+    def transform_args(*args, **kwargs):
+        copied_positional_args = list(args)
+        http_request = copied_positional_args[1]
+
+        transform_request(http_request, recording_id)
+
+        return tuple(copied_positional_args), kwargs
+
+    def combined_call(*args, **kwargs):
+        adjusted_args, adjusted_kwargs = transform_args(*args, **kwargs)
+        result = original_transport_func(*adjusted_args, **adjusted_kwargs)
+
+        # make the x-recording-upstream-base-uri the URL of the request
+        # this makes the request look like it was made to the original endpoint instead of to the proxy
+        # without this, things like LROPollers can get broken by polling the wrong endpoint
+        parsed_result = url_parse.urlparse(result.request.url)
+        upstream_uri = url_parse.urlparse(result.request.headers["x-recording-upstream-base-uri"])
+        upstream_uri_dict = {"scheme": upstream_uri.scheme, "netloc": upstream_uri.netloc}
+        original_target = parsed_result._replace(**upstream_uri_dict).geturl()
+
+        result.request.url = original_target
+        return result
+
+    RequestsTransport.send = combined_call
+
+    # store info pertinent to the test in a dictionary that other fixtures can access
+    variable_recorder = VariableRecorder(variables)
+    test_info = {"test_id": test_id, "variables": variable_recorder}
+    yield test_info  # yield and allow test to run
+
+    RequestsTransport.send = original_transport_func  # test finished running -- tear down
+
+    if hasattr(request.node, "test_error"):
+        # Exceptions are logged here instead of being raised because of how pytest handles error raising from inside
+        # fixtures and hooks. Raising from a fixture raises an error in addition to the test failure report, and the
+        # test proxy error is logged before the test failure output (making it difficult to find in pytest output).
+        # Raising from a hook isn't allowed, and produces an internal error that disrupts test execution.
+        # ResourceNotFoundErrors during playback indicate a recording mismatch
+        error = request.node.test_error
+        if isinstance(error, ResourceNotFoundError):
+            error_body = ContentDecodePolicy.deserialize_from_http_generics(error.response)
+            message = error_body.get("message") or error_body.get("Message")
+            logger = logging.getLogger()
+            logger.error(f"\n\n-----Test proxy playback error:-----\n\n{message}")
+
+    stop_record_or_playback(test_id, recording_id, variables)
+
+
+@pytest.fixture
+def variable_recorder(recorded_test) -> "dict[str, Any]":
```

Comment:
This is very cool. by referencing fixture `recorded_test` here, does this always ensure that this fixture will use the output of the previous fixture? Does order of fixture calls matter? EG:

`def test_func(variable_recorder, recorded_test)`

Reply:
The `recorded_test` reference makes it so that tests that want to use recorded variables only have to reference this `variable_recorder` fixture to get full recording functionality 🙂

I'm pretty sure both fixtures can be requested by a test and that it shouldn't affect functionality (I think pytest constructs an order to run each fixture exactly once in, similar to how class inheritance is worked out). Verifying that assumption is part of my remaining work to get this ready for review!